crates/invoker-impl/src/invocation_task.rs

```diff
@@ -409,46 +405,146 @@ where
+                    )));
+
+                // todo: We should support resuming an invocation with a newer protocol version if
+                //  the endpoint supports it
+                if !ServiceProtocolVersion::is_supported(chosen_deployment.service_protocol_version)
+                {
+                    shortcircuit!(Err(InvocationTaskError::UnsupportedServiceProtocolVersion(
+                        chosen_deployment.service_protocol_version
+                    )));
+                }
+
+                (
+                    deployment_metadata,
+                    chosen_deployment.service_protocol_version,
+                    /* has_changed= */ false,
+                )
             } else {
                 // We can choose the freshest deployment for the latest revision
                 // of the registered service.
                 let deployment = shortcircuit!(self
                     .deployment_metadata_resolver
                     .resolve_latest_deployment_for_service(self.invocation_target.service_name())
                     .ok_or(InvocationTaskError::NoDeploymentForService));
-                (deployment, /* has_changed= */ true)
+
+                let chosen_service_protocol_version =
+                    shortcircuit!(ServiceProtocolVersion::choose_max_supported_version(
+                        &deployment.metadata.supported_protocol_versions,
+                    )
+                    .ok_or_else(|| {
+                        InvocationTaskError::IncompatibleServiceEndpoint(
+                            deployment.id,
+                            deployment.metadata.supported_protocol_versions.clone(),
+                        )
+                    }));
+
+                (
+                    deployment,
+                    chosen_service_protocol_version,
+                    /* has_changed= */ true,
+                )
             };
 
-        self.send_invoker_tx(InvocationTaskOutputInner::SelectedDeployment(
-            deployment.id,
+        self.send_invoker_tx(InvocationTaskOutputInner::ChosenDeployment(
+            ChosenDeployment::new(deployment.id, chosen_service_protocol_version),
             deployment_changed,
         ));
 
+        // create a versioned connection
+        let invocation_connection =
+            VersionedInvocationConnection::new(self, chosen_service_protocol_version);
+
+        invocation_connection
+            .run(journal_metadata, deployment, journal_stream, state_iter)
+            .await
+    }
+}
+
+impl<SR, JR, EE, DMR> InvocationTask<SR, JR, EE, DMR> {
+    fn send_invoker_tx(&self, invocation_task_output_inner: InvocationTaskOutputInner) {
+        let _ = self.invoker_tx.send(InvocationTaskOutput {
+            partition: self.partition,
+            invocation_id: self.invocation_id,
+            inner: invocation_task_output_inner,
+        });
+    }
+}
+
+struct VersionedInvocationConnection<'a, SR, JR, EE, DMR> {
```

Comment:
Connection is a bad name, in this context means something else. Just name it Task, i think it's fine, or you can use Stream too.

Reply:
What is the meaning of connection in this context?

Reply:
Tbf, I am also not super happy with connection as a name.

Reply:
Connection == TCP connection here, in particular once we start doing pooling.

Reply:
Alrighty, I will find another name.